torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -98,6 +98,50 @@ struct TORCH_CUDA_API BinaryOp : public Expr {
   Val* const rhs_;
 };
 
+/*
+ * A specialization for Ternary operations.
+ * There are 3 inputs and 1 output
+ * Examples include:
+ *  1) Threshold
+ *  2) Where
+ */
+struct TORCH_CUDA_API TernaryOp : public Expr {
+  ~TernaryOp() = default;
+  TernaryOp(TernaryOpType _type, Val* _out, Val* _in1, Val* _in2, Val* _in3);
+
+  TernaryOp(const TernaryOp& other) = delete;
+  TernaryOp& operator=(const TernaryOp& other) = delete;
+
+  TernaryOp(TernaryOp&& other) = delete;
+  TernaryOp& operator=(TernaryOp&& other) = delete;
+
+  Val* out() const noexcept {
+    return out_;
+  }
+  Val* in1() const noexcept {
```

Comment:
Oops, looks like we are not being consistent there. `Expr` inherits from `IRInputOutput` which provides a pair of `deque<Val*>` for `inputs_` and `outputs_`. But that is not used by any of the operations defined in `ir_internal_nodes.h`.

We need to fix this @kevinstephano

Reply:
NVM, I didn't understand they are just helper pointers. let's leave it as is.

I'll fix the api use. Sorry for the false alarm.

Reply:
Generally speaking I think we could make a mechanical Operation class that is perhaps templatized  to cover what seems like very repetitive and mechanical code instead of having specific classes based on the number of inputs.  The additional data member pointers do seem a little redundant now that you point them out.

Perhaps the question is whether to clean up now or later...